File: 0938-Range_Sum_of_BST/Range_Sum_of_BST.java
class Solution {
    int sum = 0;
    public int rangeSumBST(TreeNode root, int low, int high) {
        binarySearch(root, low, high);
        return sum;
    }
    
    private void binarySearch(TreeNode root, int low, int high) {
        if (root == null) return;
        if (root.val < low){
            binarySearch(root.right, low, high);
        } else if (root.val > high) {
            binarySearch(root.left, low, high);
        } else {
            sum += root.val;
            binarySearch(root.left, low, high);
            binarySearch(root.right, low, high);
        }
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
  }
}

File: 0938-Range_Sum_of_BST/Range_Sum_of_BST.go
package rangesumofbst

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rangeSumBST(root *TreeNode, low int, high int) int {
	return binarySearch(root, low, high)
}

func binarySearch(root *TreeNode, low int, high int) int {
	if root == nil {
		return 0
	}
	if root.Val < low {
		return binarySearch(root.Right, low, high)
	} else if root.Val > high {
		return binarySearch(root.Left, low, high)
	} else {
		return root.Val + binarySearch(root.Left, low, high) + binarySearch(root.Right, low, high)
	}

}
